Panic during message expiration

Here is the stack from an user:
```
panic: failed to switch state from fileInUse to fileOpened for file "<dir>/msgs.52.dat", state=4
goroutine 6462 [running]:
github.com/nats-io/nats-streaming-server/stores.(*filesManager).unlockFile(...)
        /home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:1108
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).doLockFiles(0xc001e54a80, 0xc000514b90, 0xc000514b01, 0x0, 0x700000000000000)
        /home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2217 +0x32e
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).lockIndexFile(...)
        /home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2255
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).expireMsgs(0xc001e54a80, 0x1636019e86726fc1, 0x2260ff9290000, 0x0)
        /home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2906 +0x455
github.com/nats-io/nats-streaming-server/stores.(*FileMsgStore).backgroundTasks(0xc001e54a80)
        /home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:3268 +0x67b
created by github.com/nats-io/nats-streaming-server/stores.(*FileStore).newFileMsgStore
        /home/travis/gopath/src/github.com/nats-io/nats-streaming-server/stores/filestore.go:2116 +0x893
```

--- a/stores/filestore.go
+++ b/stores/filestore.go
@@ -2213,8 +2213,12 @@ func (ms *FileMsgStore) doLockFiles(fslice *fileSlice, onlyIndexFile bool) error
 	}
 	idxWasOpened, err = ms.fm.lockFile(fslice.idxFile)
 	if err != nil {
-		if !datWasOpened {
-			ms.fm.unlockFile(fslice.file)
+		if !onlyIndexFile {
+			if !datWasOpened {
+				ms.fm.closeLockedFile(fslice.file)
+			} else {
+				ms.fm.unlockFile(fslice.file)
+			}
 		}
 		return err
 	}

--- a/stores/filestore_msg_test.go
+++ b/stores/filestore_msg_test.go
@@ -1328,6 +1328,34 @@ func TestFSPanicOnStoreCloseWhileMsgsExpire(t *testing.T) {
 	fs.Close()
 }
 
+func TestFSPanicOnMsgExpireWithClosedDatFile(t *testing.T) {
+	cleanupFSDatastore(t)
+	defer cleanupFSDatastore(t)
+
+	limits := testDefaultStoreLimits
+	limits.MaxAge = 500 * time.Millisecond
+
+	fs, _ := newFileStore(t, testFSDefaultDatastore, &limits, SliceConfig(1, 0, 0, ""))
+	defer fs.Close()
+
+	cs := storeCreateChannel(t, fs, "foo")
+	for i := 0; i < 3; i++ {
+		storeMsg(t, cs, "foo", uint64(i+1), []byte("msg"))
+	}
+
+	ms := cs.Msgs.(*FileMsgStore)
+	ms.Lock()
+	idxFileName := ms.files[1].idxFile.name
+	os.Remove(idxFileName)
+	err := ioutil.WriteFile(idxFileName, []byte("xxxxxxx"), 0666)
+	ms.Unlock()
+	if err != nil {
+		t.Fatalf("Error rewriting index file: %v", err)
+	}
+
+	time.Sleep(750 * time.Millisecond)
+}
+
 func TestFSMsgIndexFileWithExtraZeros(t *testing.T) {
 	cleanupFSDatastore(t)
 	defer cleanupFSDatastore(t)

